pyg_lib/csrc/random/cpu/biased_sampling.cpp

```diff
@@ -0,0 +1,166 @@
+                          scalar_t* cdf) {
+  at::parallel_for(0, rowptr_size - 1, at::internal::GRAIN_SIZE / rowptr_size,
+                   [&](int64_t _s, int64_t _e) {
+                     // CDF conversion for each row
+                     for (int64_t i = _s; i < _e; i++) {
+                       const scalar_t* beg = bias + rowptr_data[i];
+                       int64_t len = rowptr_data[i + 1] - rowptr_data[i];
+                       scalar_t* out_beg = cdf + rowptr_data[i];
+
+                       // Remember sum and current element to
+                       // enable the in-place option (bias == cdf).
+                       scalar_t sum = 0;
+
+                       for (int64_t j = 0; j < len; j++) {
+                         sum += beg[j];
+                       }
+
+                       scalar_t cur = sum;
+                       for (int64_t j = len - 1; j >= 0; j--) {
+                         cur -= beg[j];
+                         out_beg[j] = cur / sum;
+                       }
+                     }
+                   });
+}
+
+std::pair<at::Tensor, at::Tensor> biased_to_alias(at::Tensor rowptr,
+                                                  at::Tensor bias) {
+  TORCH_CHECK(rowptr.is_cpu(), "'rowptr' must be a CPU tensor");
+  TORCH_CHECK(bias.is_cpu(), "'bias' must be a CPU tensor");
+
+  at::Tensor alias = at::empty_like(bias, rowptr.options());
+  at::Tensor out_bias = at::empty_like(bias);
+
+  int64_t rowptr_size = rowptr.size(0);
+  int64_t* rowptr_data = rowptr.data_ptr<int64_t>();
+  int64_t* alias_data = alias.data_ptr<int64_t>();
+
+  AT_DISPATCH_FLOATING_TYPES(bias.scalar_type(), "biased_to_cdf_inplace", [&] {
+    scalar_t* bias_data = bias.data_ptr<scalar_t>();
+    scalar_t* out_bias_data = out_bias.data_ptr<scalar_t>();
+    biased_to_alias_helper(rowptr_data, rowptr_size, bias_data, out_bias_data,
+                           alias_data);
+  });
+
+  return {out_bias, alias};
+}
+
+template <typename scalar_t>
+void biased_to_alias_helper(int64_t* rowptr_data,
+                            int64_t rowptr_size,
+                            const scalar_t* bias,
+                            scalar_t* out_bias,
+                            int64_t* alias) {
+  scalar_t eps = 1e-6;
+  at::parallel_for(
+      0, rowptr_size - 1, at::internal::GRAIN_SIZE / rowptr_size,
+      [&](int64_t _s, int64_t _e) {
+        // Calculate the average bias
+        for (int64_t i = _s; i < _e; i++) {
+          const scalar_t* beg = bias + rowptr_data[i];
+          int64_t len = rowptr_data[i + 1] - rowptr_data[i];
+          scalar_t* out_beg = out_bias + rowptr_data[i];
+          int64_t* alias_beg = alias + rowptr_data[i];
+          scalar_t avg = 0;
+
+          for (int64_t j = 0; j < len; j++) {
+            avg += beg[j];
+          }
+          avg /= len;
+
+          // The sets for index with a bias lower or higher than average
+          std::vector<std::pair<int64_t, scalar_t>> high, low;
```

Comment:
One more optimization we can do is to use `std::vector<size_t> high, low` instead of pair.
Later you can do
```
scalar_t low_bias = out_beg[low_idx];
out_beg[low_idx] = low_bias / avg;
out_beg[high_idx] -= (avg - low_bias);
```
Create pairs are much more costly than just use int.